polaris-migrator/src/migrations/replace-sass-motion/replace-sass-motion.ts

```diff
@@ -1,33 +1,144 @@
+      if (decl[processed]) return;
+
+      const handlers = {
+        transition: handleTransitionProps,
+        'transition-duration': handleTransitionProps,
+        'transition-delay': handleTransitionProps,
+      };
+
+      if (!isKeyOf(handlers, decl.prop)) return;
+
+      /**
+       * A collection of transformable values to migrate (e.g. `decl` lengths, functions, etc.)
+       *
+       * Note: This is evaluated at the end of each visitor execution to determine whether
+       * or not to replace the declaration or insert a comment.
+       */
+      const targets: {replaced: boolean}[] = [];
+      let hasNumericOperator = false;
+
+      const parsedValue = valueParser(decl.value);
+
+      handlers[decl.prop]();
+
+      if (targets.some(({replaced}) => !replaced || hasNumericOperator)) {
+        decl.before(postcss.comment({text: POLARIS_MIGRATOR_COMMENT}));
+        decl.before(
+          postcss.comment({
+            text: `${decl.prop}: ${parsedValue.toString()};`,
+          }),
+        );
+      } else {
+        decl.value = parsedValue.toString();
+      }
+
+      //
+      // Handlers
+      //
+
+      function handleTransitionProps() {
+        parsedValue.walk((node) => {
+          if (node.type === 'word') {
+            if (globalValues.has(node.value)) return;
+
+            if (isNumericOperator(node)) {
+              hasNumericOperator = true;
+              return;
+            }
+
+            const dimension = valueParser.unit(node.value);
+
+            if (!isTransformableDuration(dimension)) return;
+
+            targets.push({replaced: false});
+
+            const duration = `${dimension.number}${dimension.unit}`;
+
+            if (!isKeyOf(durationMap, duration)) return;
+
+            targets.at(-1)!.replaced = true;
+
+            node.value = `var(${durationMap[duration]})`;
+
+            return;
+          }
+
+          if (node.type === 'function') {
+            if (isSassFunction(namespacedDuration, node)) {
+              targets.push({replaced: false});
+
+              const args = getFunctionArgs(node);
+
+              if (!(args.length === 0 || args.length === 1)) return;
+
+              // `duration()` args reference:
+              // https://github.com/Shopify/polaris/blob/2b14c0b60097f75d21df7eaa744dfaf84f8f53f7/documentation/guides/legacy-polaris-v8-public-api.scss#L679
+              const variant = args[0]?.replace(/['"]/g, '') ?? 'base';
+
+              if (!isKeyOf(durationFunctionMap, variant)) return;
+
+              const durationCustomProperty = durationFunctionMap[variant];
+
+              targets.at(-1)!.replaced = true;
+
+              node.value = 'var';
+              node.nodes = [
+                {
+                  type: 'word',
+                  value: durationCustomProperty,
+                  sourceIndex: node.nodes[0]?.sourceIndex ?? 0,
+                  sourceEndIndex: durationCustomProperty.length,
+                },
+              ];
```

Comment:
Similar to my comment above about `replaceNodeWithValue` being designed to handle this. It also feels inconsistent with the approach used earlier where it just sets `node.value = \`var(${durationMap[duration]})\`;`.